[WIP] UpdateLog tests

wokalski opened the pull request:
This introduces `UpdateLog` tests. `UpdateLog` is a data structure which contains a description of mutations to the instance tree in order to get to the desired state.

The most important part is `subTreeChange` which highlights the type of subtree change. 
- `Nested` means that the underlying subtree did change but it was an element deep down the tree.
- `NoChange` means that the tree has not changed.
- `ReplaceElements` means that a subtree should be replaced
- `PrependElement` means that an element should be prepended

**Todos:**

- [ ] Improve formatting in tests. I want to handle this first since poor readability affects productivity a lot
- [x] Determine why the existing test is failing :)
- [x] Implement tests which verify the `UpdateLog` step by step. It's important since we can determine that the minimal amount of steps is indeed in the `UpdateLog`.
- [x] ~~Test `applyUpdateLog`. The tree from `applyUpdateLog` should be the same as newly returned element.~~
- [x] Introduce side effects in `applyUpdateLog`
- [ ] `applyUpdateLog` should perform mutations to the tree and invoke render-cycle lifecycle functions. In order to test that we should build a list of all mutations and lifecycle function invocations and assert if they happen. Similar to what we do with `UpdateLog`. The `RenderUpdateLog` however will only be a part of tests. Also, in order to test it properly we need to introduce mutations to the `TestRenderer` because currently the `hostView` is immutable.
- [ ] add tests for `prependElement`
- [ ] Remember to include `updateInstance` in the tests

wokalski committed:
First tests for UpdateLog

wokalski committed:
[No tests] Make subTreeChanged describe the type of change

Determining the subtree change at the reconciliation time lets us
perform minimal changes to the underlying views.

It also introduces formatting testInstances as JSX

rauanmayemir committed:
restructure dirs, module names, jbuild files

rauanmayemir committed:
custom alcotest check; better formatting of the output

rauanmayemir committed:
custom alcotest check; better formatting of the output

rauanmayemir committed:
added subtree change comparison func, fixed top level update tests

wokalski commented on test/Test.re:
```diff
@@ -170,6 +170,15 @@ module TestRenderer = {
+      compareElement(left, right)
+    | (ReplaceElements(left1, left2), ReplaceElements(right1, right2)) =>
+      compareElement(left1, right1) && compareElement(left2, right2)
+    | (_, _) => false;
```
👍

wokalski commented on test/Assert.re:
```diff
@@ -0,0 +1,68 @@
+  show_line(msg);
+  if (! equal(t, x, y)) {
+    Fmt.strf(
+      "Error %s:\n\nEXPECTED:@\n\n%a\n\nACTUAL:@\n\n%a\n",
```
I see, this is why you decided to recreate `Assert`?

rauanmayemir replied:
Correct. `check` doesn't allow changing the formatter string and all those helper methods are private. I'll think of something else later.

wokalski reviewed (commented):
Everything looks nice so far.

wokalski committed:
[Failing} Replace NewRenderedElement with TopLevelUpdate

This commit also includes some changes in the algorithm to ignore
empty updates.

wokalski committed:
Remove setProps and layout from nativeElement

wokalski committed:
Make updateable components stateful

rauanmayemir committed:
more restructuring

wokalski commented on test/TestRenderer.re:
```diff
@@ -0,0 +1,272 @@
+    }
+  );
+
+let printElement = formatter => Fmt.pf(formatter, "%a", printTreeFormatter());
```
I'd extract printers to a separate module, too.

wokalski commented on test/TestComponents.re:
```diff
@@ -0,0 +1,90 @@
+module Box = {
```
Maybe we should colocate these components with the main components. Maybe Inside `module Test` in a given component module? We could do `Div.Test`. On the flip side it's worse because we cannot simply `open TestComponents`

rauanmayemir committed:
factor out test printers

wokalski committed:
Remove the distinction between NativeInstance and Instance

rauanmayemir committed:
test all UpdateLog items on subtree change

wokalski commented on test/Test.re:
```diff
@@ -448,46 +463,41 @@ let suite =
             rendered0,
             Nested(
               "div",
               [ReasonReact.stringToElement("before"), Nested("div", [box_])]
```
We need to assert that the rendered `box_` instances here are physically equal (`===`) for `rendered0` and `rendered1`

rauanmayemir committed:
more updateLog tests for reduced components

rauanmayemir committed:
added SwitchComponent update log, topLevelUpdate refactoring [wip]

wokalski committed:
[Tests failing] Make topLevelUpdate compile

rauanmayemir committed:
fixed test renderer to handle new topLevelUpdate log

wokalski commented on test/Test.re:
```diff
@@ -625,21 +636,26 @@ let suite =
+              TestRenderer.{
+                typ: Nested,
+                updateLog:
+                  ref([
```
It shouldn't be ref here. Probably neither should it be ref when returned from `update`

wokalski commented on src/lib/ReactCore.rei:
```diff
@@ -180,7 +179,7 @@ module Make:
      * Log of operations performed to update an instance tree.
      */
-    module UpdateLog: {type t; let create: unit => t;};
+    module UpdateLog: {type t; type topLevelUpdate; let create: unit => t;};
```
I think that `topLevelUpdate` type should be colocated with the `update` function since it's specific to it.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -11,7 +13,7 @@ module Make = (Implementation: HostImplementation) => {
     let reset = () => {
       debug := true;
       componentKeyCounter := 0;
-      instanceIdCounter := 1 /* id 0 reserved for defaultStringInstance */
```
I don't think we have `defaultStringInstance`. We can remove this comment and set `instanceIdCounter` to 0.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -76,9 +78,8 @@ module Make = (Implementation: HostImplementation) => {
     | Nested(string, list(reactElement))
   and nativeElement = {
     make: int => Implementation.hostView,
```
`make` should be `unit => Implementation.hostView` and there should be `Implementation.memoizeInstance: (int, Implementation.hostView) => unit`.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -164,97 +164,139 @@ module Make = (Implementation: HostImplementation) => {
-      componentChanged: bool,
-      stateChanged: bool,
-      subTreeChanged: bool
+      subTreeChanged: subtreeChange
```
subTreeChanged is always replaceElements here, right?

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -164,97 +164,139 @@ module Make = (Implementation: HostImplementation) => {
-      | NewRenderedElement(renderedElement);
+      | SwitchComponent(switchComponent);
     type t = ref(list(entry));
+    type topLevelUpdate = {
```
I think this should be extracted to whatever module it's used in.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -164,97 +164,139 @@ module Make = (Implementation: HostImplementation) => {
+      subTreeChanged: subtreeChange
     };
     type entry =
       | UpdateInstance(update)
```
`SwitchComponent` => `ChangeComponent`?

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -164,97 +164,139 @@ module Make = (Implementation: HostImplementation) => {
+                        instance('state, 'action, 'elementType),
+                        instance('state, 'action, 'elementType)
+                      ): instanceUpdate;
     type update = {
```
I think this should be `instanceUpdate`. My line of thinking is that an entry is sort of a command (verb) `Update the Instance` and it contains a payload (a noun) `instanceUpdate`.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -164,97 +164,139 @@ module Make = (Implementation: HostImplementation) => {
-      let iState = component.initialState();
-      let id = GlobalState.instanceIdCounter^;
-      incr(GlobalState.instanceIdCounter);
+    let createInitialSelf = (~component) => {
```
This should probably increment the global `id` counter and return the `id`. Now the mutation is in two places but always colocated to this call.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -164,97 +164,139 @@ module Make = (Implementation: HostImplementation) => {
+        ~state=instance.iState,
+        ~pendingStateUpdates=instance.pendingStateUpdates
+      );
+    let createInstance =
```
Maybe this function is redundant and we should just create a record wherever it's used now?

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -337,32 +378,44 @@ module Make = (Implementation: HostImplementation) => {
-          component.didMount(self)
+          component.didMount(self);
         };
         switch component.elementType {
```
We need to extract this branching somewhere. It's completely redundant and there's only one difference.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -403,64 +462,53 @@ module Make = (Implementation: HostImplementation) => {
+        let Instance({element}) = opaqueInstance;
+        stateNotUpdated && element === nextElement;
       };
       let logUpdate =
```
I believe these should be extracted to the `UpdateLog` module.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -403,64 +462,53 @@ module Make = (Implementation: HostImplementation) => {
-            component.handedOffInstance := Some(updatedInstance);
-            (() => component.handedOffInstance := None)
-          };
+        let resetHandedOffInstance = {
```
I'd remove this function and just do  `component.handedOffInstance := None` inside the switch.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -403,64 +462,53 @@ module Make = (Implementation: HostImplementation) => {
           resetHandedOffInstance();
-          logString("Updating " ++ nextComponent.debugName);
           let instance = {
             ...updatedInstance,
```
Naming here is crucial to fix.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -478,39 +526,48 @@ module Make = (Implementation: HostImplementation) => {
+              newOpaqueInstance,
+              subTreeChanged
+            ) =
               switch nextComponent.elementType {
```
Another redundant branching.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -575,65 +617,67 @@ module Make = (Implementation: HostImplementation) => {
-          let self = createSelf(~instance=nextInstance);
+          let Instance({component: {debugName}, instanceSubTree, subElements}) = updatedOpaqueInstance;
+          let id = GlobalState.instanceIdCounter^;
+          incr(GlobalState.instanceIdCounter);
```
Unless I'm completely wrong, there are different `id` only in `SwitchComponent`

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -575,65 +617,67 @@ module Make = (Implementation: HostImplementation) => {
+           */
+          let (nextSubtree, newOpaqueInstance) =
             switch nextComponent.elementType {
             | React =>
```
YARB (Yet another redundant branching)

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -730,19 +800,39 @@ module Make = (Implementation: HostImplementation) => {
+            newOpaqueInstancesAndUpdates
+          );
+        let newOpaqueInstances =
+          List.map(((_, x)) => x, newOpaqueInstancesAndUpdates);
```
Use `snd` instead of that.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -812,23 +894,26 @@ module Make = (Implementation: HostImplementation) => {
     /***
      * Flush the pending updates in an instance tree.
      * TODO: invoke lifecycles
+     * TODO: (Maybe) return topLevelChange from here
```
No, `topLevelChange` should not be returned from here since the changes are happening within the tree.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -899,75 +977,65 @@ module Make = (Implementation: HostImplementation) => {
         let view =
           switch (Implementation.getInstance(id)) {
           | Some(x) => x
           | None => nativeElement.make(id)
```
```
let hostView = nativeElement.make();
Implementation.memoizeView(id, hostView);
hostView;
```

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -899,75 +977,65 @@ module Make = (Implementation: HostImplementation) => {
         fromOpaqueInstance(nearestParentView),
         Render.flattenRenderedElement(renderedElement)
       );
+    /**
```
Let's remove this comment. This information is in some comment above already.

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -1058,30 +1088,28 @@ module Make = (Implementation: HostImplementation) => {
+      );
     };
   };
   module LayoutTest = {
```
We need to throw this out of here. `Layout` is no longer tightly coupled.

wokalski commented on test/ReasonReact.re:
```diff
@@ -0,0 +1,44 @@
+include ReactCore_Internal.Make(Implementation);
+
+module Text = {
+  /**
```
Fixed.

wokalski reviewed (commented):
I've found some things to fix in the first part of our session.

rauanmayemir committed:
general refactoring, types renaming, functions clean up

wokalski commented on src/lib/ReactCore_Internal.re:
```diff
@@ -451,163 +440,106 @@ module Make = (Implementation: HostImplementation) => {
-          resetHandedOffInstance();
-          let instance = {
+          updatedComponent.handedOffInstance := None;
+          let updatedInstance = {
```
This is `updatedInstanceWithNewElement`.

rauanmayemir committed:
make tests compile

wokalski committed:
Fix incorrect oldInstance passed to UpdateInstance

wokalski committed:
Fix element comparators in tests

wokalski committed:
Remove old and new Subtree from ChangeComponent

They are included inside new/old instance

wokalski committed:
Remove Layout from ReactCore

wokalski committed:
Remove id from nativeElement.make.

Instances are now memoized by the library

rauanmayemir committed:
revert test renderer changes

wokalski committed:
Implement applyUpdateLog to mutate the baking tree

wokalski committed:
[Buggy] Brings rendering on iOS back

wokalski commented:
`jbuilder build -x ios` should work but it doesn't.

In order to build currently you need to do:
```
jbuilder build -x ios
ocamlfind -toolchain ios ocamlopt -I ./_build/default.ios/src/lib -I ./_build/default.ios/src/flex/.flex.objs -I ./_build/default.ios/src/lib/.r2n2.objs -I ./_build/default.ios/src/flex -package unix -linkpkg -output-complete-obj flex.cmxa r2n2.cmxa -impl ./_build/default.ios/src/bin/react_ios.re.ml -o react_ios.o
```
and then proceed with the Xcode build.

rauanmayemir committed:
render log tests [wip]

render log tests [wip]

rauanmayemir committed:
implemented render log tests, added test comparator ppxs

wokalski committed:
Add RenderLog tests for Change counter test
